UniversalAlgebra/Complexity/CSP.lagda

```diff
@@ -15,67 +15,83 @@ module Complexity.CSP where
-  * 𝑅 is an 𝑛-ary relation on 𝐷, called the constraint relation of 𝐶.
+ 1. a scope function,  s : I → var, and
 
-\begin{code}
+ 2. a constraint relation, i.e., a predicate over the function type I → D
+
+        I ···> var
+         .     .
+          .   .
+           ⌟ ⌞
+            D
 
-private variable χ ρ : Level
 
-record Constraint (∣V∣ ∣D∣ : ℕ) : Type (lsuc (χ ⊔ ρ)) where
- field
-  scope : Fin ∣V∣ → Type χ -- a subset of Fin ∣V∣
+The *scope* of a constraint is an indexed subset of the set of variable symbols.
+We could define a type for this, e.g.,
 
-  relation : ((v : Fin ∣V∣ ) → v ∈ scope → Fin ∣D∣) → Type ρ
-  -- a subset of "tuples" (i.e., functions 𝑓 : scope → Fin ∣D∣)
+```
+ Scope : Type ν → Type ι → _
+ Scope V I = I → V
+```
 
- -- An assignment 𝑓 : 𝑉 → 𝐷 of values in the domain to variables in 𝑉
- -- *satisfies* the constraint 𝐶 = (𝑥, 𝑅) if 𝑓 ↾ 𝑥 ∈ 𝑅.
- satisfies : (Fin ∣V∣ → Fin ∣D∣) → Type ρ
- satisfies f = (λ v _ → f v) ∈ relation
+but we omit this definition because it's so simple; to reiterate,
+a scope of "arity" I on "variables" V is simply a map from I to V,
+where,
 
+* I denotes the "number" of variables involved in the scope
+* V denotes a collection (type) of "variable symbols"
+
+\begin{code}
 
-open Constraint
+module _ -- levels for...
+         {ι : Level} -- ...arity (or argument index) types
+         {ν : Level} -- ...variable symbol types
+         {δ : Level} -- ... domain types
+         -- {ρ : Level} -- ... relation types
+         where
 
-record CSP : Type (lsuc (χ ⊔ ρ)) where
- field
-  ∣V∣ -- # of variables,
-   ∣D∣ -- # of elements in the domain,
-    ∣C∣ -- # of constraints (resp.)
-       : ℕ
-  𝐶    : Fin ∣C∣ → Constraint{χ} ∣V∣ ∣D∣
+ record Constraint (var : Type ν)(dom : Type δ){ρ : Level} : Type (lsuc (ι ⊔ ν ⊔ δ ⊔ ρ)) where
+  field
+   arity  : Type ι               -- The "number" of variables involved in the constraint.
+   scope  : arity → var          -- Which variables are involved in the constraint.
+   rel    : Rel dom arity {ρ}    -- The constraint relation.
 
- -- An assignment 𝑓 *solves* the instance 𝑃 if it satisfies all the constraints.
- isSolution : (Fin ∣V∣ → Fin ∣D∣) → Type ρ
- isSolution f = ∀ i → (satisfies (𝐶 i)) f
+  satisfies : (var → dom) → Type ρ  -- An assignment 𝑓 : var → dom of values to variables
+  satisfies f = f ∘ scope ∈ rel     -- *satisfies* the constraint 𝐶 = (σ , 𝑅) provided
+                                    -- 𝑓 ∘ σ ∈ 𝑅, where σ is the scope of the constraint.
 
 \end{code}
 
+#### CSP Instances
 
+An instance of a constraint satisfaction problem is a triple 𝑃 = (𝑉, 𝐷, 𝐶) where
 
-#### References
+* 𝑉 denotes a set of "variables"
+* 𝐷 denotes a "domain",
+* 𝐶 denotes an indexed collection of constraints.
 
-Some of the informal (i.e., non-agda) material in this module is similar to that presented in
+\begin{code}
+
+ record CSPInstance  (var : Type ν)(dom : Type δ){ρ : Level} : Type (lsuc (ι ⊔ ν ⊔ δ ⊔ ρ)) where
```

Comment:
This seems oddly formed to me - shouldn't the `Constraint` be a parameter?

Reply:
I think I was not precise. Isn't a `CSPInstance` supposed to be an instance of a CSP? In other words, it should be parametrized by a single "problem". The definition used here only have a very diffuse 'problem' that it depends on, quite indirectly.